File: services/tracking/src/features/tracking/domain/status.py
"""Tracking statuses and the forward-only state machine (JE-89).

PURE domain code: no database, no HTTP, no framework imports. Both callers of the
state machine depend on that purity —

  * `PUT /v1/trackings/{orderId}/status`, the carrier endpoint, which turns a
    rejection into a `400 Bad Request`; and
  * TestMode automatic progression, which runs on a 10s timer with no request
    context at all.

so nothing here may assume an HTTP request, a session, or a caller identity.
"""

from dataclasses import dataclass
from enum import StrEnum
from typing import Final


class TrackingStatus(StrEnum):
    """The four valid delivery statuses.

    A `StrEnum` so the member compares and serializes as its own name — matching
    both the `VARCHAR(50)` storage and the REST surface, which takes and returns it
    as a string. One representation across wire, storage and HTTP.
    """

    SHIPPED = "SHIPPED"
    ON_THE_WAY = "ON_THE_WAY"
    OUT_FOR_DELIVERY = "OUT_FOR_DELIVERY"
    DELIVERED = "DELIVERED"


# The allowed progression, in order. Position in this tuple IS the ordering the
# guards below compare against — `TrackingStatus` is a StrEnum, so its members
# compare alphabetically, NOT in delivery order (`DELIVERED` < `SHIPPED`
# alphabetically). Never order statuses by comparing the enum members directly.
STATUS_ORDER: Final[tuple[TrackingStatus, ...]] = (
    TrackingStatus.SHIPPED,
    TrackingStatus.ON_THE_WAY,
    TrackingStatus.OUT_FOR_DELIVERY,
    TrackingStatus.DELIVERED,
)

#: The status every tracking is created at.
INITIAL_STATUS: Final[TrackingStatus] = STATUS_ORDER[0]

#: Terminal status. Nothing may follow it, and nothing may update a tracking on it.
TERMINAL_STATUS: Final[TrackingStatus] = STATUS_ORDER[-1]


class TransitionRejectionReason(StrEnum):
    """Why a transition was rejected.

    Three distinct members on purpose. A single `new > current` comparison would
    satisfy all three guards at once and collapse them into one indistinguishable
    failure — this enum keeps "already delivered", "went backwards" and "did not
    move" separately identifiable, both for the `reason` field the logging
    convention requires on `*_failed` events and for the tests that must verify
    each guard independently.
    """

    #: Guard 1 — the current status is DELIVERED, which is terminal.
    ALREADY_DELIVERED = "already_delivered"
    #: Guard 2 — the requested status is earlier in the progression.
    BACKWARD_TRANSITION = "backward_transition"
    #: Guard 3 — the requested status equals the current one; not strictly forward.
    NOT_STRICTLY_FORWARD = "not_strictly_forward"


class InvalidTransitionError(Exception):
    """Raised by `assert_can_transition` when a transition is rejected.

    Carries the machine-readable `reason` so the HTTP layer can log it (and, if it
    chooses, surface it) without re-deriving why the transition failed.
    """

    def __init__(
        self,
        current: TrackingStatus,
        requested: TrackingStatus,
        reason: TransitionRejectionReason,
    ) -> None:
        self.current = current
        self.requested = requested
        self.reason = reason
        super().__init__(
            f"cannot transition from {current} to {requested}: {reason}"
        )


@dataclass(frozen=True, slots=True)
class TransitionCheck:
    """Result of evaluating a transition. Falsy-safe via `allowed`."""

    allowed: bool
    reason: TransitionRejectionReason | None = None


def status_index(status: TrackingStatus) -> int:
    """Return `status`'s position in the forward-only progression."""
    return STATUS_ORDER.index(status)


def check_transition(
    current: TrackingStatus, requested: TrackingStatus
) -> TransitionCheck:
    """Evaluate `current -> requested` against the three guards.

    The guards are checked in this order, and the order is load-bearing: a
    `DELIVERED -> SHIPPED` request violates guards 1 and 2 simultaneously, and
    `DELIVERED -> DELIVERED` violates 1 and 3. Terminality is the more specific
    fact about the tracking, so it is reported first.

    1. `DELIVERED` is terminal — ANY update against it is rejected.
    2. No backward transitions — the requested status may not be earlier.
    3. Strictly forward — a status EQUAL to the current one is also rejected.
    """
    # Guard 1: terminal. Checked before the ordering guards so that a tracking
    # already delivered reports "already_delivered" whatever is requested of it,
    # including DELIVERED itself.
    if current is TERMINAL_STATUS:
        return TransitionCheck(False, TransitionRejectionReason.ALREADY_DELIVERED)

    current_index = status_index(current)
    requested_index = status_index(requested)

    # Guard 2: backward.
    if requested_index < current_index:
        return TransitionCheck(False, TransitionRejectionReason.BACKWARD_TRANSITION)

    # Guard 3: strictly forward — equal is not forward. Distinct from guard 2:
    # guard 2 fires on `<`, this one only on `==`, so the two can never both be
    # the reported reason for the same pair.
    if requested_index == current_index:
        return TransitionCheck(False, TransitionRejectionReason.NOT_STRICTLY_FORWARD)

    return TransitionCheck(True)


def can_transition(current: TrackingStatus, requested: TrackingStatus) -> bool:
    """True when `current -> requested` is allowed.

    Thin boolean view of `check_transition`.
    """
    return check_transition(current, requested).allowed


def assert_can_transition(
    current: TrackingStatus, requested: TrackingStatus
) -> None:
    """Raise `InvalidTransitionError` unless `current -> requested` is allowed.

    The form callers use when a rejection is exceptional — the PUT handler maps
    the raised error to `400 Bad Request`.
    """
    result = check_transition(current, requested)
    if not result.allowed:
        # `reason` is always set when `allowed` is False.
        assert result.reason is not None
        raise InvalidTransitionError(current, requested, result.reason)


def next_status(current: TrackingStatus) -> TrackingStatus | None:
    """Return the single status that follows `current`, or None at the terminal one.

    Used by TestMode progression (Phase E) to advance one step every 10 seconds.
    Deliberately returns None rather than raising at `DELIVERED`: reaching the end
    of the progression is the expected way a TestMode run finishes, not an error.
    """
    if current is TERMINAL_STATUS:
        return None
    return STATUS_ORDER[status_index(current) + 1]


def parse_status(value: str) -> TrackingStatus:
    """Parse an external string into a `TrackingStatus`.

    Raises `ValueError` for anything outside the four valid values — the REST
    handler turns that into a `400`. Case-sensitive on purpose: the four values
    are a fixed wire contract shared with the proto, not free-form input.
    """
    try:
        return TrackingStatus(value)
    except ValueError as exc:
        valid = ", ".join(s.value for s in STATUS_ORDER)
        raise ValueError(
            f"invalid tracking status {value!r}; expected one of: {valid}"
        ) from exc

File: services/tracking-go/internal/domain/status.go
package domain

import "fmt"

// Status is one of the five delivery statuses.
//
// Underlying type string so it stores directly into VARCHAR(50) and serializes
// directly onto the REST surface — one representation across wire, storage and
// HTTP. Stored as a plain VARCHAR rather than a MySQL ENUM: the REST surface
// carries it as a string, and widening a native ENUM is a DDL change.
type Status string

const (
	StatusPlaced         Status = "PLACED"
	StatusProcessing     Status = "PROCESSING"
	StatusShipped        Status = "SHIPPED"
	StatusOutForDelivery Status = "OUT_FOR_DELIVERY"
	StatusDelivered      Status = "DELIVERED"
)

// statusOrder is the allowed progression, in order. Position in this slice IS
// the ordering every guard below compares against.
//
// NEVER order statuses by comparing the Status values directly. Status has
// underlying type string, so `<` compiles and silently yields ALPHABETICAL
// order, in which DELIVERED < PLACED — the terminal status sorting before the
// initial one. That comparison would type-check, run, and be wrong. The index
// lookup below is the only ordering in this package.
var statusOrder = [...]Status{
	StatusPlaced,
	StatusProcessing,
	StatusShipped,
	StatusOutForDelivery,
	StatusDelivered,
}

const (
	// InitialStatus is the status every tracking is created at.
	InitialStatus = StatusPlaced
	// TerminalStatus is terminal: nothing may follow it, and nothing may update
	// a tracking sitting on it.
	TerminalStatus = StatusDelivered
)

// RejectionReason is the machine-readable reason a transition was rejected.
//
// Three distinct values on purpose. A single `requested > current` comparison
// would satisfy all three guards at once and collapse them into one
// indistinguishable failure. Keeping them separate gives the `reason` field the
// logging convention requires on *_failed events, and lets each guard be tested
// independently.
type RejectionReason string

const (
	// ReasonAlreadyDelivered — guard 1: the current status is terminal.
	ReasonAlreadyDelivered RejectionReason = "already_delivered"
	// ReasonBackwardTransition — guard 2: the requested status is earlier.
	ReasonBackwardTransition RejectionReason = "backward_transition"
	// ReasonNotStrictlyForward — guard 3: the requested status equals the current.
	ReasonNotStrictlyForward RejectionReason = "not_strictly_forward"
)

// TransitionCheck is the result of evaluating a transition. Reason is "" when
// Allowed is true.
type TransitionCheck struct {
	Allowed bool
	Reason  RejectionReason
}

// InvalidTransitionError is returned by AssertCanTransition. It carries the
// machine-readable Reason so the HTTP layer can log it without re-deriving why
// the transition failed.
type InvalidTransitionError struct {
	Current   Status
	Requested Status
	Reason    RejectionReason
}

func (e *InvalidTransitionError) Error() string {
	return fmt.Sprintf("cannot transition from %s to %s: %s", e.Current, e.Requested, e.Reason)
}

// StatusIndex returns s's position in the forward-only progression. ok is false
// for a value outside the five.
func StatusIndex(s Status) (int, bool) {
	for i, candidate := range statusOrder {
		if candidate == s {
			return i, true
		}
	}
	return 0, false
}

// CheckTransition evaluates current -> requested against the three guards.
//
// The guards run in this order and THE ORDER IS LOAD-BEARING. DELIVERED->PLACED
// violates guards 1 and 2 simultaneously; DELIVERED->DELIVERED violates 1 and 3.
// Terminality is the more specific fact about the tracking, so it is reported
// first.
//
// Skipping is ALLOWED: PLACED -> DELIVERED is legal. This is a forward-only
// machine, not a next-step-only one.
func CheckTransition(current, requested Status) TransitionCheck {
	// Guard 1: terminal. Checked before the ordering guards so that a tracking
	// already delivered reports already_delivered whatever is requested of it,
	// INCLUDING DELIVERED itself.
	if current == TerminalStatus {
		return TransitionCheck{Allowed: false, Reason: ReasonAlreadyDelivered}
	}

	currentIndex, ok := StatusIndex(current)
	if !ok {
		return TransitionCheck{Allowed: false, Reason: ReasonBackwardTransition}
	}
	requestedIndex, ok := StatusIndex(requested)
	if !ok {
		return TransitionCheck{Allowed: false, Reason: ReasonBackwardTransition}
	}

	// Guard 2: backward.
	if requestedIndex < currentIndex {
		return TransitionCheck{Allowed: false, Reason: ReasonBackwardTransition}
	}

	// Guard 3: strictly forward — equal is not forward. Distinct from guard 2:
	// guard 2 fires on `<`, this one only on `==`, so the two can never both be
	// the reported reason for the same pair.
	if requestedIndex == currentIndex {
		return TransitionCheck{Allowed: false, Reason: ReasonNotStrictlyForward}
	}

	return TransitionCheck{Allowed: true}
}

// CanTransition is the boolean view of CheckTransition.
func CanTransition(current, requested Status) bool {
	return CheckTransition(current, requested).Allowed
}

// AssertCanTransition returns an *InvalidTransitionError unless the transition
// is allowed. The form callers use when a rejection is exceptional — the PUT
// handler maps the returned error to 400 Bad Request.
func AssertCanTransition(current, requested Status) error {
	result := CheckTransition(current, requested)
	if result.Allowed {
		return nil
	}
	return &InvalidTransitionError{
		Current:   current,
		Requested: requested,
		Reason:    result.Reason,
	}
}

// NextStatus returns the single status following current.
//
// ok is false at the terminal status. Deliberately not an error: reaching the
// end of the progression is the expected way a TestMode run finishes.
func NextStatus(current Status) (Status, bool) {
	if current == TerminalStatus {
		return "", false
	}
	index, ok := StatusIndex(current)
	if !ok {
		return "", false
	}
	return statusOrder[index+1], true
}

// ParseStatus parses an external string into a Status.
//
// Case-SENSITIVE on purpose: the five values are a fixed wire contract shared
// with the proto, not free-form input. The REST handler turns the returned error
// into a 400.
func ParseStatus(s string) (Status, error) {
	for _, candidate := range statusOrder {
		if string(candidate) == s {
			return candidate, nil
		}
	}
	return "", fmt.Errorf(
		"invalid tracking status '%s'; expected one of: %s, %s, %s, %s, %s",
		s,
		StatusPlaced, StatusProcessing, StatusShipped, StatusOutForDelivery, StatusDelivered,
	)
}
